### ISSUE TITLE
Results are not correctly written to `out` in dask.array functions

### ISSUE THREAD

**Issue by JDWarner** [ContinuumIO/dask#429]
Many NumPy functions allow an optional arg/kwarg `out` for results to be written into. When this feature is used in `dask.array`, the result returned by the function is correct but the dask array passed to `out` does not contain correct values.

I've tested and confirmed this for `da.maximum`, `da.minimum`, `da.cos`, and `da.sin` after which I decided it was likely a general problem.

Minimal example (replace the two `result = ` calls to test different functions):

``` python
import numpy as np
import dask.array as da

test = da.arange(-128, 128, chunks=(128,))
out = da.empty(256, chunks=(128,))
result = da.minimum(test, 0, out=out)  # result is correct, out is not

nptest = test.compute()
npout = np.empty(256)
npresult = np.minimum(nptest, 0, out=npout)  # stored correctly in out

# Test NumPy first
np.testing.assert_allclose(npresult, npout)  # succeeds
np.testing.assert_allclose(result.compute(), out.compute())  # fails
```

EDIT: NumPy describes the intended behavior of `out` as: "If `out` is provided, the function writes the result into it, and returns a reference to `out`."

**Comment by mrocklin** [comment]
Support for `out` is indeed lacking.  It's unfortunate that this didn't err.  Many `dask.array` functions accept `kwargs` and so silently fail in cases like this.

There are two solutions

1.  Support out parameters
2.  Improve handling of keyword arguments

I think that the choice between these two depends on whether or not out-parameters are valuable enough to warrant the extra maintenance costs.

@JDWarner why were you using out parameters?  Was it for computational efficiency or for programming convenience?

Computational efficiency in this case won't be impacted.  Using the `.store` method will store the result as it is computed.

**Comment by JDWarner** [comment]
I'll freely acknowledge that `out` is essentially irrelevant for efficiency _when using Dask_.

The reason I ran into this is because I've been making a proof-of-concept showing it's possible to support Dask arrays and NumPy arrays on a shared codebase in scikit-image. Our algorithms not infrequently use `out` to improve efficiency for NumPy, and since the option seemed to be understood I used it.

I can handle this in the compatibility layer, but in case it wasn't intended I raised the Issue.
